src/main/java/com/amazon/opendistroforelasticsearch/sql/parser/subquery/InSubqueryRewriter.java

```diff
@@ -0,0 +1,146 @@
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.parser.subquery;
+
+import com.alibaba.druid.sql.ast.SQLExpr;
+import com.alibaba.druid.sql.ast.expr.SQLBinaryOpExpr;
+import com.alibaba.druid.sql.ast.expr.SQLBinaryOperator;
+import com.alibaba.druid.sql.ast.expr.SQLInSubQueryExpr;
+import com.alibaba.druid.sql.ast.expr.SQLNullExpr;
+import com.alibaba.druid.sql.ast.statement.SQLJoinTableSource;
+import com.alibaba.druid.sql.ast.statement.SQLJoinTableSource.JoinType;
+import com.alibaba.druid.sql.dialect.mysql.ast.statement.MySqlSelectQueryBlock;
+import com.amazon.opendistroforelasticsearch.sql.parser.subquery.model.Subquery;
+
+/**
+ * IN Subquery Rewriter.
+ * For example,
+ *  SELECT * FROM A WHERE a IN (SELECT b FROM B) and c > 10 should be rewritten to
+ *  SELECT A.* FROM A JOIN B ON A.a = B.b WHERE c > 10 and B.b IS NOT NULL.
+ */
+public class InSubqueryRewriter implements SubqueryRewriter {
+    private final SQLInSubQueryExpr inSubQueryExpr;
+    private final MySqlSelectQueryBlock inSubQuery;
+    private final SQLExpr joinLeftExpr;
+    private final SQLExpr joinRightExpr;
+
+    public InSubqueryRewriter(Subquery subquery) {
+        this.inSubQueryExpr = (SQLInSubQueryExpr) subquery.getSubQueryExpr();
+        this.inSubQuery = (MySqlSelectQueryBlock) inSubQueryExpr.getSubQuery().getQuery();
+        joinLeftExpr = inSubQueryExpr.getExpr();
+        joinRightExpr = rightExpr(inSubQuery);
+    }
+
+    /**
+     * Rewrite the subquery with JOIN clause. This API is NOT idempotent.
+     *
+     * @param query Subquery will be written.
+     */
+    @Override
+    public void rewrite(MySqlSelectQueryBlock query) {
+        query.setFrom(buildFrom(query));
+        query.setWhere(buildWhere(query));
+    }
+
+    /**
+     * Build From clause from input query.
+     *
+     * With the input query.
+     *       Query
+     *     /   |   \
+     * SELECT FROM     WHERE
+     *    |    |     /   |  \
+     *    *    A  c>10 AND INSubquery
+     *                      /    \
+     *                     a    Query
+     *                          /    \
+     *                       SELECT FORM
+     *                         |     |
+     *                         b     B
+     *
+     * The FROM logic should be
+     *     JOIN
+     *     / | \
+     *    A  B A.a == B.b
+     */
+    private SQLJoinTableSource buildFrom(MySqlSelectQueryBlock rootQuery) {
```

Comment:
np: I would call this `buildFromClause` and the other one `buildWhereClause` to avoid potential confusion with this method being a factory method for the class, which builds an instance from the argument.

Reply:
This is private method, so I don't think that there might be much of a confusion

Reply:
It's subjective, but to me naming and access are orthogonal concerns.